lib/src/metta/runner/builtin_mods/json.rs

```diff
@@ -0,0 +1,223 @@
+    let metatype = atom_to_string(&get_meta_type(&input));
+    let space = DynSpace::new(GroundingSpace::new());
+    let typ = get_atom_types(&space, &input);
+    if metatype == "Expression" {
+        let exp = TryInto::<&ExpressionAtom>::try_into(input)?;
+        match encode_list(exp.clone().into_children()) {
+            Ok(encoded_list) => Ok(encoded_list),
+            Err(err) => Err(err),
+        }
+    }
+    else if atom_to_string(&typ[0]) == "Number" {
+        let number_atom_string = atom_to_string(&input);
+        Ok(number_atom_string)
+    }
+    else if atom_to_string(&typ[0]) == "SpaceType" {
+        match encode_dictspace(&input) {
+            Ok(str) => Ok(str),
+            Err(err) => Err(err)
+        }
+    }
+    // String should be encoded using serde_json since it will encode "a" with additional escape
+    // characters. Symbols like just 'a' are not supported by serde_json so proposition is to encode
+    // them like strings.
+    else {
+        let atom_string = atom_to_string(&input);
+        match serde_json::to_string(&atom_string) {
+            Ok(encoded) => Ok(encoded),
+            Err(err) => Err(ExecError::from(format!("Encode string failed: {}", err))),
+        }
+    }
+}
+
+fn json_encode(args: &[Atom]) -> Result<Vec<Atom>, ExecError> {
+    let arg_error = || ExecError::from("json-encode expects atom to encode as input");
+    let input = args.get(0).ok_or_else(arg_error)?;
+
+    match encode_atom(input) {
+        Ok(str) => Ok(vec![Atom::gnd(Str::from_string(str))]),
+        Err(err) => Err(err)
+    }
+}
+
+fn decode_value(v: &Value) -> Result<Atom, ExecError> {
+    match v {
+        Value::Array(_) => {match decode_array(v.as_array().unwrap()) {
+            Ok(atom) => Ok(atom),
+            Err(err) => Err(err),
+        }},
+        Value::Null => Ok(Atom::gnd(Str::from_str("null"))),
+        Value::Bool(_) => Ok(Atom::gnd(Bool(v.as_bool().unwrap()))),
+        Value::String(_) => Ok(Atom::gnd(Str::from_string(v.to_string()))),
+        Value::Object(_) => {match decode_object(&v.clone()) {
+            Ok(atom) => Ok(atom),
+            Err(err) => Err(err),
+        }},
+        Value::Number(_) => {
+            if v.is_i64() {
+                Ok(Atom::gnd(Number::Integer(v.as_i64().unwrap())))
+            }
+            else if v.is_f64() {
+                Ok(Atom::gnd(Number::Float(v.as_f64().unwrap())))
+            }
+            else {
+                Err(ExecError::from("Currently u64 type is not supported"))
+            }
+        },
+    }
+}
+
+fn decode_object(obj: &Value) -> Result<Atom, ExecError> {
+    let dict = obj.as_object().unwrap();
+    let keys = dict.keys();
+    let space = DynSpace::new(GroundingSpace::new());
+    for key in keys {
+        let value = dict.get(key).unwrap();
+        match decode_value(value) {
+            Ok(decoded) => space.borrow_mut().add(Atom::expr([Atom::gnd(Str::from_string(key.clone())), decoded])),
```

Comment:
I would not wrap `GroundingSpace` into a `DynSpace` from very beginning. Instead you can fill `GroundingSpace` instance and then wrap it into a `DynSpace` before making the atom. It will make borrowing unnecessary:
```suggestion
            Ok(decoded) => space.add(Atom::expr([Atom::gnd(Str::from_string(key.clone())), decoded])),
```